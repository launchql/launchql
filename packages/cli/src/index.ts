#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});