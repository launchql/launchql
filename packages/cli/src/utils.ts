import chalk from 'chalk';

import { readAndParsePackageJson } from './package';
import { ParsedArgs } from 'minimist';

export const extractFirst = (argv: Partial<ParsedArgs>) => {
    const first = argv._?.[0];
    const newArgv = {
      ...argv,
      _: argv._?.slice(1) ?? []
    };
    return { first, newArgv };
  };

// Function to display the version information
export function displayVersion() {
    const pkg = readAndParsePackageJson();
    console.log(chalk.green(`Name: ${pkg.name}`));
    console.log(chalk.blue(`Version: ${pkg.version}`));
}

export const usageText = `
  Usage: launchql <command> [options]
  
  Commands:
    start              Start the Launchql services.
    version, -v        Display the version of the Starship Client.
  
  Configuration File:
    --config <path>       Specify the path to the configuration file containing the actual config file. Required.
                          Command-line options will override settings from this file if both are provided.
  
  Additional Help:
    $ launchql help          Display this help information.
  `;

export function displayUsage() {
    console.log(usageText);
}
