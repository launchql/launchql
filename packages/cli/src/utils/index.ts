export * from './argv';
export * from './database';
export * from './display';