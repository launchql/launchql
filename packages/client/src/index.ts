import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}