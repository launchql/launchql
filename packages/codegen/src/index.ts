import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
