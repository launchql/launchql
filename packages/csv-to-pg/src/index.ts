// @ts-nocheck
export * from './parse';
export * from './parser';
