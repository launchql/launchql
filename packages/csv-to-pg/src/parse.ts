// @ts-nocheck

import csv from 'csv-parser';
import { createReadStream, readFileSync } from 'fs';
import { safeLoad as parseYAML } from 'js-yaml';
import * as ast from 'pg-ast';
import {
  makeBoundingBox,
  makeLocation,
  getRelatedField,
  wrapValue
} from './utils';
function isNumeric(str) {
  if (typeof str === 'number') return true;
  if (typeof str !== 'string') return false; // we only process strings!
  return (
    !isNaN(str) && // use type coercion to parse the _entirety_ of the string (`parseFloat` alone does not do this)...
    !isNaN(parseFloat(str))
  ); // ...and ensure strings of whitespace fail
}

const parseJson = (value) => {
  if (typeof value === 'string') return value;
  return value && JSON.stringify(value);
};

const psqlArray = (value) => {
  if (value && value.length) {
    return `{${value.map((v) => v)}}`;
  }
};

export const parse = (path, opts) =>
  new Promise((resolve, reject) => {
    const results = [];
    createReadStream(path)
      .pipe(csv(opts))
      // TODO check if 'data' is guaranteed to have a full row,
      // if so, make a hook to use the stream properly
      .on('data', (data) => results.push(data))
      .on('error', (er) => {
        reject(er);
      })
      .on('end', () => {
        resolve(results);
      });
  });

export const readConfig = (config) => {
  let configValue;
  if (config.endsWith('.js')) {
    configValue = require(config);
  } else if (config.endsWith('json')) {
    configValue = JSON.parse(readFileSync(config, 'utf-8'));
  } else if (config.endsWith('yaml') || config.endsWith('yml')) {
    configValue = parseYAML(readFileSync(config, 'utf-8'));
  } else {
    throw new Error('unsupported config!');
  }
  return configValue;
};

const getFromValue = (from) => {
  if (Array.isArray(from)) return from;
  return [from];
};

// looks like the CSV library gives us empty strings?
const cleanseEmptyStrings = (str) => {
  if (typeof str === 'string') {
    if (str.trim() === '') return null;
    return str;
  } else {
    return str;
  }
};

const parseBoolean = (str) => {
  if (typeof str === 'boolean') {
    return str;
  } else if (typeof str === 'string') {
    const s = str.toLowerCase();
    if (s === 'true') {
      return true;
    } else if (s === 't') {
      return true;
    } else if (s === 'f') {
      return false;
    } else if (s === 'false') {
      return false;
    }
    return null;
  } else {
    return null;
  }
};

const getValuesFromKeys = (object, keys) => keys.map((key) => object[key]);

const identity = (a) => a;

const isEmpty = (value) => value === null || typeof value === 'undefined';

// type (int, text, etc)
// from Array of keys that map to records found (e.g., ['lon', 'lat'])
const getCoercionFunc = (type, from, opts) => {
  const parse = (opts.parse = opts.parse || identity);

  switch (type) {
    case 'int':
      return (record) => {
        const value = parse(record[from[0]]);
        if (isEmpty(value)) {
          return ast.Null({});
        }
        if (!isNumeric(value)) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.Integer({ ival: value })
        });
        return wrapValue(val, opts);
      };
    case 'float':
      return (record) => {
        const value = parse(record[from[0]]);
        if (isEmpty(value)) {
          return ast.Null({});
        }
        if (!isNumeric(value)) {
          return ast.Null({});
        }

        const val = ast.A_Const({
          val: ast.Float({ str: value })
        });
        return wrapValue(val, opts);
      };
    case 'boolean':
    case 'bool':
      return (record) => {
        const value = parse(parseBoolean(record[from[0]]));

        if (isEmpty(value)) {
          return ast.Null({});
        }

        const val = ast.String({
          str: value ? 'TRUE' : 'FALSE'
        });
        return wrapValue(val, opts);
      };
    case 'bbox':
      // do bbox magic with args from the fields
      return (record) => {
        const val = makeBoundingBox(parse(record[from[0]]));
        return wrapValue(val, opts);
      };
    case 'location':
      return (record) => {
        const [lon, lat] = getValuesFromKeys(record, from);
        if (typeof lon === 'undefined') {
          return ast.Null({});
        }
        if (typeof lat === 'undefined') {
          return ast.Null({});
        }
        if (!isNumeric(lon) || !isNumeric(lat)) {
          return ast.Null({});
        }

        // NO parse here...
        const val = makeLocation(lon, lat);
        return wrapValue(val, opts);
      };
    case 'related':
      return (record) => {
        return getRelatedField({
          ...opts,
          record,
          from
        });
      };
    case 'uuid':
      return (record) => {
        const value = parse(record[from[0]]);
        if (
          isEmpty(value) ||
          !/^([0-9a-fA-F]{8})-(([0-9a-fA-F]{4}-){3})([0-9a-fA-F]{12})$/i.test(
            value
          )
        ) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.String({ str: value })
        });
        return wrapValue(val, opts);
      };
    case 'text':
      return (record) => {
        const value = parse(cleanseEmptyStrings(record[from[0]]));
        if (isEmpty(value)) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.String({ str: value })
        });
        return wrapValue(val, opts);
      };

    case 'text[]':
      return (record) => {
        const value = parse(psqlArray(cleanseEmptyStrings(record[from[0]])));
        if (isEmpty(value)) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.String({ str: value })
        });
        return wrapValue(val, opts);
      };
    case 'image':
    case 'attachment':
    case 'json':
    case 'jsonb':
      return (record) => {
        const value = parse(parseJson(cleanseEmptyStrings(record[from[0]])));
        if (isEmpty(value)) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.String({ str: value })
        });
        return wrapValue(val, opts);
      };
    default:
      return (record) => {
        const value = parse(cleanseEmptyStrings(record[from[0]]));
        if (isEmpty(value)) {
          return ast.Null({});
        }
        const val = ast.A_Const({
          val: ast.String({ str: value })
        });
        return wrapValue(val, opts);
      };
  }
};

export const parseTypes = (config) => {
  return Object.entries(config.fields).reduce((m, v) => {
    let [key, value] = v;
    let type;
    let from;
    if (typeof value === 'string') {
      type = value;
      from = [key];
      if (['related', 'location'].includes(type)) {
        throw new Error('must use object for ' + type + ' type');
      }
      value = {
        type,
        from
      };
    } else {
      type = value.type;
      from = getFromValue(value.from || key);
    }
    m[key] = getCoercionFunc(type, from, value);
    return m;
  }, {});
};
