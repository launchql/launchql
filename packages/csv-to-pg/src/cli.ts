#!/usr/bin/env node
// @ts-nocheck
import { prompt } from '@pyramation/prompt';
import { readConfig } from './parse';
import { Parser } from './parser';
import { normalizePath } from './utils';
import { dirname } from 'path';
import { writeFileSync } from 'fs';

const argv = process.argv.slice(2);

(async () => {
  let { config } = await prompt(
    [
      {
        _: true,
        name: 'config',
        type: 'config',
        required: true
      }
    ],
    argv
  );

  config = normalizePath(config);
  const dir = dirname(config);
  config = readConfig(config);

  if (config.input) {
    if (!argv.includes('--input')) {
      argv.push('--input');
      config.input = normalizePath(config.input, dir);
      argv.push(config.input);
    }
  }
  if (config.output) {
    if (!argv.includes('--output')) {
      argv.push('--output');
      config.output = normalizePath(config.output, dir);
      argv.push(config.output);
    }
  }

  const results = await prompt(
    [
      {
        name: 'input',
        type: 'path',
        required: true
      },
      {
        name: 'output',
        type: 'path',
        required: true
      }
    ],
    argv
  );

  config.input = results.input;

  let outFile = results.output;
  if (!outFile.endsWith('.sql')) outFile = outFile + '.sql';

  if (argv.includes('--debug')) {
    config.debug = true;
  }

  const parser = new Parser(config);
  const sql = await parser.parse();

  writeFileSync(config.output, sql);
})();
