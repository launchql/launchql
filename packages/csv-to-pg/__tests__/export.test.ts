// @ts-nocheck

import { parse, parseTypes } from '../src';
import { resolve } from 'path';
import { deparse } from 'pgsql-deparser';
import { InsertOne, InsertMany } from '../src/utils';
import { Parser } from '../src/parser';

const testCase = resolve(__dirname + '/../__fixtures__/test-case.csv');

const config = {
  schema: 'collections_public',
  singleStmts: true,
  table: 'field',
  headers: [
    'id',
    'database_id',
    'table_id',
    'name',
    'description',
    'smart_tags',
    'is_required',
    'default_value',
    'is_hidden',
    'type',
    'field_order',
    'regexp',
    'chk',
    'chk_exp',
    'min',
    'max',
    'created_at',
    'updated_at'
  ],
  fields: {
    id: 'uuid',
    database_id: 'uuid',
    table_id: 'uuid',
    name: 'text',
    description: 'text'
  }
};

it('test case', async () => {
  const records = await parse(testCase, { headers: config.headers });
  const types = parseTypes(config);
  const stmt = InsertMany({
    schema: config.schema,
    table: config.table,
    types,
    records
  });

  expect(deparse([stmt])).toMatchSnapshot();
});

it('test case parser', async () => {
  const parser = new Parser(config);

  const sql = await parser.parse([
    {
      id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      database_id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      table_id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      name: 'name here',
      description: 'description'
    }
  ]);

  expect(sql).toMatchSnapshot();
});

it('jsonb/json', async () => {
  const parser = new Parser({
    schema: 'collections_public',
    singleStmts: true,
    table: 'field',
    fields: {
      id: 'uuid',
      name: 'text',
      data: 'jsonb'
    }
  });

  const sql = await parser.parse([
    {
      id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      name: 'name here',
      data: {
        a: 1
      }
    }
  ]);

  expect(sql).toMatchSnapshot();
});

it('image/attachment', async () => {
  const parser = new Parser({
    schema: 'collections_public',
    singleStmts: true,
    table: 'field',
    fields: {
      id: 'uuid',
      name: 'text',
      image: 'image',
      upload: 'attachment'
    }
  });

  const sql = await parser.parse([
    {
      id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      name: 'name here',
      image: {
        url: 'http://path/to/image.jpg'
      },
      upload: {
        url: 'http://path/to/image.jpg'
      }
    }
  ]);

  expect(sql).toMatchSnapshot();
});

it('arrays', async () => {
  const parser = new Parser({
    schema: 'collections_public',
    singleStmts: true,
    table: 'field',
    fields: {
      id: 'uuid',
      schemas: 'text[]'
    }
  });

  const sql = await parser.parse([
    {
      id: '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4',
      schemas: ['a', 'b']
    }
  ]);

  expect(sql).toMatchSnapshot();
});
