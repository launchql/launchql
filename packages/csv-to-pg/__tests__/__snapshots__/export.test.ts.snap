// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrays 1`] = `"INSERT INTO collections_public.field ( id, schemas ) VALUES ('450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', '{a,b}');"`;

exports[`image/attachment 1`] = `"INSERT INTO collections_public.field ( id, name, image, upload ) VALUES ('450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', 'name here', '{"url":"http://path/to/image.jpg"}', '{"url":"http://path/to/image.jpg"}');"`;

exports[`jsonb/json 1`] = `"INSERT INTO collections_public.field ( id, name, data ) VALUES ('450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', 'name here', '{"a":1}');"`;

exports[`test case 1`] = `"INSERT INTO collections_public.field ( id, database_id, table_id, name, description ) VALUES ('391f5dfe-a251-43af-bee5-821e69214d20', '95482a0f-ed8e-447d-ab05-318650d2f328', '41c74ae0-b474-4aaf-b561-faadbadfc154', 'id', NULL), ('c32175b2-fe04-496a-8725-4ee35d6b2a3d', '95482a0f-ed8e-447d-ab05-318650d2f328', '41c74ae0-b474-4aaf-b561-faadbadfc154', 'owner_id', NULL);"`;

exports[`test case parser 1`] = `"INSERT INTO collections_public.field ( id, database_id, table_id, name, description ) VALUES ('450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', '450e3b3b-b68d-4abc-990c-65cb8a1dcdb4', 'name here', 'description');"`;
